pkg/client/remotecommand/remotecommand_test.go

```diff
@@ -18,271 +18,171 @@ package remotecommand
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	s.readCalled = true
-	b := []byte(s.data)
-	n := copy(p, b)
-	return n, io.EOF
-}
-
-func (s *fakeUpgradeStream) Write(p []byte) (int, error) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	s.writeCalled = true
-	s.dataWritten = make([]byte, len(p))
-	copy(s.dataWritten, p)
-	return len(p), io.EOF
-}
-
-func (s *fakeUpgradeStream) Close() error {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	s.closeCalled = true
-	return nil
-}
-
-func (s *fakeUpgradeStream) Reset() error {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	s.resetCalled = true
-	return nil
-}
-
-func (s *fakeUpgradeStream) Headers() http.Header {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	return http.Header{}
-}
-
 func TestRequestExecuteRemoteCommand(t *testing.T) {
 	testCases := []struct {
-		Upgrader    *fakeUpgrader
-		Stdin       string
-		Stdout      string
-		Stderr      string
-		Error       string
-		Tty         bool
-		ShouldError bool
+		Stdin  string
+		Stdout string
+		Stderr string
+		Error  string
+		Tty    bool
 	}{
 		{
-			Upgrader:    &fakeUpgrader{err: errors.New("bail")},
-			ShouldError: true,
+			Error: "bail",
 		},
 		{
-			Upgrader:    &fakeUpgrader{conn: newFakeUpgradeConnection()},
-			Stdin:       "a",
-			Stdout:      "b",
-			Stderr:      "c",
-			Error:       "bail",
-			ShouldError: true,
+			Stdin:  "a",
+			Stdout: "b",
+			Stderr: "c",
 		},
 		{
-			Upgrader: &fakeUpgrader{conn: newFakeUpgradeConnection()},
-			Stdin:    "a",
-			Stdout:   "b",
-			Stderr:   "c",
-		},
-		{
-			Upgrader: &fakeUpgrader{conn: newFakeUpgradeConnection()},
-			Stdin:    "a",
-			Stdout:   "b",
-			Stderr:   "c",
-			Tty:      true,
+			Stdin:  "a",
+			Stdout: "b",
+			Tty:    true,
 		},
 	}
 
 	for i, testCase := range testCases {
-		if testCase.Error != "" {
-			testCase.Upgrader.conn.errorData = testCase.Error
-		}
-		if testCase.Stdout != "" {
-			testCase.Upgrader.conn.stdoutData = testCase.Stdout
-		}
-		if testCase.Stderr != "" {
-			testCase.Upgrader.conn.stderrData = testCase.Stderr
-		}
-		var localOut, localErr *bytes.Buffer
-		if testCase.Stdout != "" {
-			localOut = &bytes.Buffer{}
-		}
-		if testCase.Stderr != "" {
-			localErr = &bytes.Buffer{}
-		}
-		e := New(&client.Request{}, &client.Config{}, []string{"ls", "/"}, strings.NewReader(testCase.Stdin), localOut, localErr, testCase.Tty)
-		e.upgrader = testCase.Upgrader
-		err := e.Execute()
-		hasErr := err != nil
-		if hasErr != testCase.ShouldError {
-			t.Fatalf("%d: expected %t, got %t: %v", i, testCase.ShouldError, hasErr, err)
+		localOut := &bytes.Buffer{}
+		localErr := &bytes.Buffer{}
+
+		// error + stdin + stdout
+		expectedStreams := 3
+		if !testCase.Tty {
+			// stderr
+			expectedStreams++
 		}
 
-		conn := testCase.Upgrader.conn
-		if testCase.Error != "" {
-			if conn.errorStream == nil {
-				t.Fatalf("%d: expected error stream creation", i)
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
It's from `handleExec` Kubelet's server.go, modified for testing. I can definitely split it out for readability, but I'm not sure it could be shared w/the Kubelet.